src/nvim/autocmd.c

```diff
@@ -2470,45 +2481,53 @@ bool autocmd_delete_id(int64_t id)
 // ===========================================================================
 //  AucmdExecutable Functions
 // ===========================================================================
-char *aucmd_exec_default_desc(AucmdExecutable acc)
+
+/// Generate a string description of a callback
+static char *aucmd_callback_to_string(Callback cb)
 {
-  size_t msglen = 100;
+  // NOTE: this function probably belongs in a helper
 
-  switch (acc.type) {
-  case CALLABLE_CB:
-    switch (acc.callable.cb.type) {
-    case kCallbackLua: {
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<Lua function %d>", acc.callable.cb.data.luaref);
-      return msg;
-    }
-    case kCallbackFuncref: {
-      // TODO(tjdevries): Is this enough space for this?
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<vim function: %s>", acc.callable.cb.data.funcref);
-      return msg;
-    }
-    case kCallbackPartial: {
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<vim partial: %s>", acc.callable.cb.data.partial->pt_name);
-      return msg;
-    }
-    default:
-      return NULL;
-    }
+  // FIXME: enabling dynamic memory will cause problems when called from `nvim_get_autocmds()`
```

Comment:
In `nvim_get_autocmds` change this line

```c
        PUT(autocmd_info,
            "command",
            STRING_OBJ(cstr_to_string(aucmd_exec_to_string(ac, ac->exec))));
```

to

```c
        PUT(autocmd_info,
            "command",
            STRING_OBJ(cstr_as_string(aucmd_exec_to_string(ac, ac->exec))));
```

Then the API layer will take care of freeing the alloc'd memory.

Reply:
this seems to be the missing puzzle piece 🙌 
any call to `free()` in this function was causing all sorts of memory issues, even in the cleanup section.